src/include/mpir_hwtopo.h

```diff
@@ -0,0 +1,114 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#ifndef MPIR_HWTOPO_H_INCLUDED
+#define MPIR_HWTOPO_H_INCLUDED
+
+/* hwtopo types */
+typedef enum {
+    MPIR_HWTOPO_TYPE__NONE = -1,
+    MPIR_HWTOPO_TYPE__NODE,
+    MPIR_HWTOPO_TYPE__PACKAGE,
+    MPIR_HWTOPO_TYPE__SOCKET,
+    MPIR_HWTOPO_TYPE__CPU,
+    MPIR_HWTOPO_TYPE__CORE,
+    MPIR_HWTOPO_TYPE__HWTHREAD,
+    MPIR_HWTOPO_TYPE__L1CACHE,
+    MPIR_HWTOPO_TYPE__L2CACHE,
+    MPIR_HWTOPO_TYPE__L3CACHE,
+    MPIR_HWTOPO_TYPE__L4CACHE,
+    MPIR_HWTOPO_TYPE__L5CACHE,
+    MPIR_HWTOPO_TYPE__DDR,
+    MPIR_HWTOPO_TYPE__HBM,
+    MPIR_HWTOPO_TYPE__PCI_DEVICE,
+    MPIR_HWTOPO_TYPE__MAX
+} MPIR_hwtopo_type_e;
+
+#define MPIR_HWTOPO_GID_INVALID (-1)
+
+typedef int MPIR_hwtopo_gid_t;
+
+/* Topology managing APIs start here. These allow for the
+ * users to to init/finalize and query the status of the
+ * topology layer during runtime */
+
+/*
+ * Initialize hw topology
+ */
+int MPII_hwtopo_init(void);
+
+/*
+ * Finalize hw topology
+ */
+int MPII_hwtopo_finalize(void);
+
+/*
+ * Check whether hw topology is initialized or not
+ */
+bool MPIR_hwtopo_is_initialized(void);
+
+
+/* Topology tree APIs start here. These allow for users to
+ * get an entry point in the topology tree, usually this is
+ * the lowest object in the tree that has matching cpuset
+ * with the current process */
+
+/*
+ * Return the unique gid of the leaf object for this process
+ */
+MPIR_hwtopo_gid_t MPIR_hwtopo_get_leaf(void);
+
+/*
+ * Return the depth in the topology tree for the given gid
+ */
+int MPIR_hwtopo_get_depth(MPIR_hwtopo_gid_t gid);
+
+/*
+ * Return the gid of the ancestor at depth level
+ */
+MPIR_hwtopo_gid_t MPIR_hwtopo_get_ancestor(MPIR_hwtopo_gid_t gid, int depth);
```

Comment:
Again, I think we would rather see an example than stating the obvious.

Reply:
done